19.0 tutorials joyep

yewentai opened the pull request.

yewentai committed:
[ADD] estate: Chapter 2 -  initialize estate module with manifest and init file

yewentai committed:
[ADD] estate: Chapter 3 - add estate module with models and field

yewentai committed:
[ADD] estate: Chapter 4 - add access control for estate properties

yewentai committed:
[ADD] estate: Chapter 5 - add fields, menus, and views

yewentai committed:
[ADD] estate: Chapter 6 - add form, tree, and search layouts

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1036.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1036)

yewentai committed:
[ADD] estate: Chapter 7 - add property types, tags, and offers

Add property types, tags, and offers with corresponding views and access rights

antonrom1 commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,15 @@
+              name="Properties"
+              parent="estate_property_menu_categ" 
+              action="estate_property_action"
+              sequence="20"/>
```
Good! 👍
Just as a note, you can also define them like [this](https://github.com/odoo/odoo/blob/7afe40e50d88448dd966d20f5ae7ac84d986e405/addons/account/views/account_menuitem.xml#L39-L42) instead of specifying the parent

yewentai replied:
thx! fixed.

yewentai committed:
[ADD] estate: Chapter 8 - add computed fields and views

Enhance property and offer models with computed fields and views

antonrom1 commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,84 @@
+
+    # Computed fields
+    @api.depends("living_area", "garden_area")
+    def _compute_total_area(self):
+        for record in self:
+            record.total_area = record.living_area + record.garden_area
+
+    total_area = fields.Integer("Total Area (sqm)", compute="_compute_total_area", store=True)
+
+    @api.depends("offer_ids.price")
+    def _compute_best_offer(self):
+        for record in self:
+            prices = record.offer_ids.mapped("price")
+            record.best_offer = max(prices) if prices else 0.0
+
+    best_offer = fields.Float("Best Offer", compute="_compute_best_offer", store=True)
+
+    @api.onchange("garden")
+    def _onchange_garden(self):
+        if not self.garden:
+            self.garden_area = 0
+            self.garden_orientation = False
+        else:
+            if not self.garden_area:
+                self.garden_area = 10
+            if not self.garden_orientation:
+                self.garden_orientation = "north"
```
[In a Model attribute order should be](https://www.odoo.com/documentation/19.0/contributing/development/coding_guidelines.html#:~:text=In%20a%20Model%20attribute%20order%20should%20be)
- Private attributes (_name, _description, _inherit, …)
- Default method and default_get
- Field declarations
- SQL constraints and indexes
- Compute, inverse and search methods in the same order as field declaration
- Selection method (methods used to return computed values for selection fields)
- Constrains methods (@api.constrains) and onchange methods (@api.onchange)
- CRUD methods (ORM overrides)
- Action methods
- And finally, other business methods.

yewentai replied:
thx! fixed.

yewentai committed:
[ADD] estate: Chapter 9 - add buttons and correspoding actions

Implement action methods for property status  and offer status

antonrom1 commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,59 @@
+                    ("property_id", "=", offer.property_id.id),
+                    ("id", "!=", offer.id),
+                ]
+            )
```
it's good, but you can make it a bit simpler. Recordsets have bitwise operations (`__sub__`, `__or__`, `__and__` etc)
```suggestion
            other_offers = self.property_id.offer_ids - offer
```

Also check that none other offers were already accepted. e.g.: 
```py
if any(other_offers.filtered(lambda o: o.status == 'accepted')): ...
```

yewentai replied:
thx! fixed.

antonrom1 commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,98 @@
+from odoo import api, fields, models
+from odoo.exceptions import UserError
+from dateutil.relativedelta import relativedelta
```
nitpick: import order https://www.odoo.com/documentation/19.0/contributing/development/coding_guidelines.html#imports

```suggestion
from dateutil.relativedelta import relativedelta

from odoo import api, fields, models
from odoo.exceptions import UserError
```

yewentai replied:
thx! fixed.

antonrom1 commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,98 @@
+    sequence = fields.Integer("Sequence", default=10)
+
+    # Basic fields
+    description = fields.Text("Description")
+    postcode = fields.Char("Postcode")
+    date_availability = fields.Date(
+        "Available From", default=lambda self: fields.Date.today() + relativedelta(months=3), copy=False
+    )
```
Don't forget the trailing comma. Also making this multiline would look better imo

yewentai replied:
thx!

antonrom1 commented on estate/models/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from . import estate_property
+from . import estate_property_type
+from . import estate_property_tag
+from . import estate_property_offer
```
nit: alphabetical import order

yewentai replied:
thx! fixed.

antonrom1 reviewed (commented):
Look really good overall, good job :P

yewentai committed:
[ADD] estate: Chapter 10 - add SQL and python constraints

Add SQL constraints for price and uniqueness in property models and add Python constraints for the accepted price.

yewentai committed:
[ADD] estate: Chapter 1 - debuger and formatter config

yewentai committed:
[ADD] estate: Chapter 11 - add the sprinkles

Enhance property management - update launch configuration, modify ordering, and improve views

yewentai committed:
[ADD] estate: Chapter 12 - add inheritances

Implement inherited user model and views, add CRUD validation for property offers

yewentai committed:
[ADD] estate: Chapter 13 - add a link module.

Initial implementation of estate module with property management features

yewentai committed:
[ADD] estate: Chapter 14 - add kanban view for properties.

yewentai committed:
[REFINE] estate: Chapter 15 - update import statements

yewentai committed:
[REFINE] awesome_owl: 1. Displaying a counter

update module structure and enhance Playground component

yewentai committed:
[REFINE] awesome modules: remove unnecessary UTF-8 encoding declarations
